File: typescript/src/mfa.ts
export const AUTHENTICATION_FACTORS = [
  'password',
  'magic-link',
  'email-otp',
  'sms-otp',
  'totp',
  'recovery-code',
  'oauth',
  'webauthn',
] as const;

export type AuthenticationFactor = (typeof AUTHENTICATION_FACTORS)[number];

export const MFA_CHALLENGE_FACTORS = [
  'email-otp',
  'sms-otp',
  'totp',
  'recovery-code',
  'webauthn',
] as const satisfies readonly AuthenticationFactor[];

export type MfaChallengeFactor = (typeof MFA_CHALLENGE_FACTORS)[number];

export const AUTHENTICATION_ASSURANCE_LEVELS = [
  'unauthenticated',
  'single-factor',
  'multi-factor',
  'phishing-resistant',
] as const;

export type AuthenticationAssuranceLevel = (typeof AUTHENTICATION_ASSURANCE_LEVELS)[number];

export interface AuthenticationEvidence {
  readonly factor: AuthenticationFactor;
  readonly verifiedAtMs: number;
  /** Whether a WebAuthn ceremony performed authenticator user verification. */
  readonly userVerified?: boolean;
}

export interface AuthenticationAssurance {
  readonly level: AuthenticationAssuranceLevel;
  /** Time at which the evidence needed for this level was most recently complete. */
  readonly authenticatedAtMs: number | null;
  readonly factorCount: number;
}

export interface StepUpPolicy {
  readonly minimumLevel: AuthenticationAssuranceLevel;
  readonly maxAgeMs?: number;
  readonly requiredFactors?: readonly AuthenticationFactor[];
}

export type StepUpChallengeReason =
  | 'unauthenticated'
  | 'insufficient-assurance'
  | 'required-factor-missing'
  | 'stale';

export type StepUpDecision =
  | {
      readonly kind: 'allow';
      readonly assurance: AuthenticationAssurance;
    }
  | {
      readonly kind: 'challenge';
      readonly reason: StepUpChallengeReason;
      readonly assurance: AuthenticationAssurance;
    };

export interface MfaChallengeFacts {
  readonly purpose: string | null;
  readonly subject: string | null;
  readonly factor: MfaChallengeFactor | null;
  readonly credentialVersion: number | null;
  readonly expiresAtMs?: number | null;
  readonly consumedAtMs?: number | null;
}

export interface MfaChallengePolicy {
  readonly allowedPurposes: readonly string[];
  readonly allowedFactors: readonly MfaChallengeFactor[];
  readonly expectedSubject?: string;
  readonly currentCredentialVersion?: number;
  readonly nowMs?: number;
}

export type MfaChallengeRejectionReason =
  | 'malformed'
  | 'purpose-mismatch'
  | 'factor-not-allowed'
  | 'subject-mismatch'
  | 'credential-version-mismatch'
  | 'already-consumed'
  | 'expired';

export type MfaChallengeDecision =
  | { readonly kind: 'accept' }
  | {
      readonly kind: 'reject';
      readonly reason: MfaChallengeRejectionReason;
    };

const assuranceRank: Readonly<Record<AuthenticationAssuranceLevel, number>> = {
  unauthenticated: 0,
  'single-factor': 1,
  'multi-factor': 2,
  'phishing-resistant': 3,
};

function isAuthenticationFactor(value: string): value is AuthenticationFactor {
  return (AUTHENTICATION_FACTORS as readonly string[]).includes(value);
}

function requireTimestamp(value: number, label: string): void {
  if (!Number.isSafeInteger(value)) {
    throw new RangeError(`${label} must be a safe integer timestamp`);
  }
}

function latestEvidenceByFactor(
  evidence: readonly AuthenticationEvidence[],
): ReadonlyMap<AuthenticationFactor, { verifiedAtMs: number; userVerifiedAtMs: number | null }> {
  const latest = new Map<
    AuthenticationFactor,
    { verifiedAtMs: number; userVerifiedAtMs: number | null }
  >();
  for (const item of evidence) {
    const factor: string = item.factor;
    if (!isAuthenticationFactor(factor)) {
      throw new RangeError(`unknown authentication factor: ${factor}`);
    }
    requireTimestamp(item.verifiedAtMs, 'verifiedAtMs');
    const current = latest.get(factor);
    const userVerifiedAtMs =
      factor === 'webauthn' && item.userVerified === true
        ? Math.max(current?.userVerifiedAtMs ?? Number.NEGATIVE_INFINITY, item.verifiedAtMs)
        : (current?.userVerifiedAtMs ?? null);
    latest.set(factor, {
      verifiedAtMs: Math.max(current?.verifiedAtMs ?? Number.NEGATIVE_INFINITY, item.verifiedAtMs),
      userVerifiedAtMs,
    });
  }
  return latest;
}

/** Derives the strongest assurance supported by the supplied verified evidence. */
export function deriveAuthenticationAssurance(
  evidence: readonly AuthenticationEvidence[],
): AuthenticationAssurance {
  const latest = latestEvidenceByFactor(evidence);
  if (latest.size === 0) {
    return {
      level: 'unauthenticated',
      authenticatedAtMs: null,
      factorCount: 0,
    };
  }

  const webauthn = latest.get('webauthn');
  if (webauthn?.userVerifiedAtMs != null) {
    return {
      level: 'phishing-resistant',
      authenticatedAtMs: webauthn.userVerifiedAtMs,
      factorCount: latest.size,
    };
  }

  const timestamps = [...latest.values()]
    .map((item) => item.verifiedAtMs)
    .sort((left, right) => right - left);
  if (timestamps.length >= 2) {
    return {
      level: 'multi-factor',
      authenticatedAtMs: timestamps[1] ?? null,
      factorCount: latest.size,
    };
  }

  return {
    level: 'single-factor',
    authenticatedAtMs: timestamps[0] ?? null,
    factorCount: 1,
  };
}

/** Determines whether existing authentication evidence satisfies a step-up policy. */
export function evaluateStepUp(
  evidence: readonly AuthenticationEvidence[],
  nowMs: number,
  policy: StepUpPolicy,
): StepUpDecision {
  requireTimestamp(nowMs, 'nowMs');
  const minimumLevel: string = policy.minimumLevel;
  if (!(minimumLevel in assuranceRank)) {
    throw new RangeError(`unknown assurance level: ${minimumLevel}`);
  }
  if (
    policy.maxAgeMs !== undefined &&
    (!Number.isSafeInteger(policy.maxAgeMs) || policy.maxAgeMs < 0)
  ) {
    throw new RangeError('maxAgeMs must be a non-negative safe integer');
  }

  const latest = latestEvidenceByFactor(evidence);
  const assurance = deriveAuthenticationAssurance(evidence);
  if (assurance.level === 'unauthenticated') {
    return { kind: 'challenge', reason: 'unauthenticated', assurance };
  }

  const requiredFactors = policy.requiredFactors ?? [];
  if (requiredFactors.some((factor) => !latest.has(factor))) {
    return { kind: 'challenge', reason: 'required-factor-missing', assurance };
  }
  if (assuranceRank[assurance.level] < assuranceRank[policy.minimumLevel]) {
    return { kind: 'challenge', reason: 'insufficient-assurance', assurance };
  }

  if (policy.maxAgeMs !== undefined) {
    const cutoffMs = nowMs - policy.maxAgeMs;
    requireTimestamp(cutoffMs, 'step-up cutoff');
    const assuranceIsStale =
      assurance.authenticatedAtMs === null || assurance.authenticatedAtMs < cutoffMs;
    const requiredFactorIsStale = requiredFactors.some(
      (factor) => (latest.get(factor)?.verifiedAtMs ?? Number.NEGATIVE_INFINITY) < cutoffMs,
    );
    if (assuranceIsStale || requiredFactorIsStale) {
      return { kind: 'challenge', reason: 'stale', assurance };
    }
  }

  return { kind: 'allow', assurance };
}

/** Selects the first enrolled challenge factor in application-supplied preference order. */
export function selectMfaChallengeFactor(
  enrolledFactors: readonly MfaChallengeFactor[],
  preferredFactors: readonly MfaChallengeFactor[],
): MfaChallengeFactor | null {
  const enrolled = new Set(enrolledFactors);
  return preferredFactors.find((factor) => enrolled.has(factor)) ?? null;
}

/** Evaluates already-decoded MFA challenge claims without owning token cryptography. */
export function evaluateMfaChallenge(
  facts: MfaChallengeFacts,
  policy: MfaChallengePolicy,
): MfaChallengeDecision {
  if (policy.allowedPurposes.length === 0 || policy.allowedFactors.length === 0) {
    throw new RangeError('MFA challenge policy must allow a purpose and factor');
  }
  if (
    policy.currentCredentialVersion !== undefined &&
    (!Number.isSafeInteger(policy.currentCredentialVersion) ||
      policy.currentCredentialVersion < 0)
  ) {
    throw new RangeError('currentCredentialVersion must be a non-negative safe integer');
  }
  if (
    facts.purpose === null ||
    facts.purpose.length === 0 ||
    facts.subject === null ||
    facts.subject.length === 0 ||
    facts.factor === null ||
    facts.credentialVersion === null ||
    !Number.isSafeInteger(facts.credentialVersion) ||
    facts.credentialVersion < 0
  ) {
    return { kind: 'reject', reason: 'malformed' };
  }
  if (!policy.allowedPurposes.includes(facts.purpose)) {
    return { kind: 'reject', reason: 'purpose-mismatch' };
  }
  if (!policy.allowedFactors.includes(facts.factor)) {
    return { kind: 'reject', reason: 'factor-not-allowed' };
  }
  if (policy.expectedSubject !== undefined && facts.subject !== policy.expectedSubject) {
    return { kind: 'reject', reason: 'subject-mismatch' };
  }
  if (
    policy.currentCredentialVersion !== undefined &&
    facts.credentialVersion !== policy.currentCredentialVersion
  ) {
    return { kind: 'reject', reason: 'credential-version-mismatch' };
  }
  if (facts.consumedAtMs != null) {
    requireTimestamp(facts.consumedAtMs, 'consumedAtMs');
    return { kind: 'reject', reason: 'already-consumed' };
  }
  if (facts.expiresAtMs != null) {
    requireTimestamp(facts.expiresAtMs, 'expiresAtMs');
    if (policy.nowMs === undefined) {
      throw new RangeError('nowMs is required when expiresAtMs is present');
    }
    requireTimestamp(policy.nowMs, 'nowMs');
    if (policy.nowMs >= facts.expiresAtMs) {
      return { kind: 'reject', reason: 'expired' };
    }
  }

  return { kind: 'accept' };
}

File: rust/src/mfa.rs
use crate::{AuthError, AuthResult};

#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub enum AuthenticationFactor {
    Password,
    MagicLink,
    EmailOtp,
    SmsOtp,
    Totp,
    RecoveryCode,
    Oauth,
    Webauthn,
}

#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub enum MfaChallengeFactor {
    EmailOtp,
    SmsOtp,
    Totp,
    RecoveryCode,
    Webauthn,
}

#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub enum AuthenticationAssuranceLevel {
    Unauthenticated,
    SingleFactor,
    MultiFactor,
    PhishingResistant,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct AuthenticationEvidence {
    pub factor: AuthenticationFactor,
    pub verified_at_ms: i64,
    pub user_verified: bool,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct AuthenticationAssurance {
    pub level: AuthenticationAssuranceLevel,
    pub authenticated_at_ms: Option<i64>,
    pub factor_count: usize,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct StepUpPolicy<'a> {
    pub minimum_level: AuthenticationAssuranceLevel,
    pub max_age_ms: Option<u64>,
    pub required_factors: &'a [AuthenticationFactor],
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum StepUpChallengeReason {
    Unauthenticated,
    InsufficientAssurance,
    RequiredFactorMissing,
    Stale,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum StepUpDecision {
    Allow {
        assurance: AuthenticationAssurance,
    },
    Challenge {
        reason: StepUpChallengeReason,
        assurance: AuthenticationAssurance,
    },
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct MfaChallengeFacts<'a> {
    pub purpose: Option<&'a str>,
    pub subject: Option<&'a str>,
    pub factor: Option<MfaChallengeFactor>,
    pub credential_version: Option<u64>,
    pub expires_at_ms: Option<i64>,
    pub consumed_at_ms: Option<i64>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct MfaChallengePolicy<'a> {
    pub allowed_purposes: &'a [&'a str],
    pub allowed_factors: &'a [MfaChallengeFactor],
    pub expected_subject: Option<&'a str>,
    pub current_credential_version: Option<u64>,
    pub now_ms: Option<i64>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum MfaChallengeRejectionReason {
    Malformed,
    PurposeMismatch,
    FactorNotAllowed,
    SubjectMismatch,
    CredentialVersionMismatch,
    AlreadyConsumed,
    Expired,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum MfaChallengeDecision {
    Accept,
    Reject(MfaChallengeRejectionReason),
}

const FACTOR_COUNT: usize = 8;

const fn factor_index(factor: AuthenticationFactor) -> usize {
    match factor {
        AuthenticationFactor::Password => 0,
        AuthenticationFactor::MagicLink => 1,
        AuthenticationFactor::EmailOtp => 2,
        AuthenticationFactor::SmsOtp => 3,
        AuthenticationFactor::Totp => 4,
        AuthenticationFactor::RecoveryCode => 5,
        AuthenticationFactor::Oauth => 6,
        AuthenticationFactor::Webauthn => 7,
    }
}

fn latest_evidence(evidence: &[AuthenticationEvidence]) -> ([Option<i64>; FACTOR_COUNT], bool) {
    let mut latest = [None; FACTOR_COUNT];
    let mut latest_webauthn_user_verified_at = None;
    for item in evidence {
        let index = factor_index(item.factor);
        if latest[index].is_none_or(|current| item.verified_at_ms > current) {
            latest[index] = Some(item.verified_at_ms);
        }
        if item.factor == AuthenticationFactor::Webauthn && item.user_verified {
            latest_webauthn_user_verified_at = Some(
                latest_webauthn_user_verified_at.map_or(item.verified_at_ms, |current: i64| {
                    current.max(item.verified_at_ms)
                }),
            );
        }
    }
    (latest, latest_webauthn_user_verified_at.is_some())
}

/// Derives the strongest assurance supported by the supplied verified evidence.
#[must_use]
pub fn derive_authentication_assurance(
    evidence: &[AuthenticationEvidence],
) -> AuthenticationAssurance {
    let (latest, webauthn_user_verified) = latest_evidence(evidence);
    let mut timestamps = latest.iter().flatten().copied().collect::<Vec<_>>();
    timestamps.sort_unstable_by(|left, right| right.cmp(left));
    if timestamps.is_empty() {
        return AuthenticationAssurance {
            level: AuthenticationAssuranceLevel::Unauthenticated,
            authenticated_at_ms: None,
            factor_count: 0,
        };
    }

    if webauthn_user_verified {
        return AuthenticationAssurance {
            level: AuthenticationAssuranceLevel::PhishingResistant,
            authenticated_at_ms: evidence
                .iter()
                .filter(|item| item.factor == AuthenticationFactor::Webauthn && item.user_verified)
                .map(|item| item.verified_at_ms)
                .max(),
            factor_count: timestamps.len(),
        };
    }

    if timestamps.len() >= 2 {
        return AuthenticationAssurance {
            level: AuthenticationAssuranceLevel::MultiFactor,
            authenticated_at_ms: timestamps.get(1).copied(),
            factor_count: timestamps.len(),
        };
    }

    AuthenticationAssurance {
        level: AuthenticationAssuranceLevel::SingleFactor,
        authenticated_at_ms: timestamps.first().copied(),
        factor_count: 1,
    }
}

/// Determines whether existing authentication evidence satisfies a step-up policy.
///
/// # Errors
///
/// Returns [`AuthError::ArithmeticOverflow`] when the freshness cutoff cannot be represented.
pub fn evaluate_step_up(
    evidence: &[AuthenticationEvidence],
    now_ms: i64,
    policy: StepUpPolicy<'_>,
) -> AuthResult<StepUpDecision> {
    let (latest, _) = latest_evidence(evidence);
    let assurance = derive_authentication_assurance(evidence);
    if assurance.level == AuthenticationAssuranceLevel::Unauthenticated {
        return Ok(StepUpDecision::Challenge {
            reason: StepUpChallengeReason::Unauthenticated,
            assurance,
        });
    }

    if policy
        .required_factors
        .iter()
        .any(|factor| latest[factor_index(*factor)].is_none())
    {
        return Ok(StepUpDecision::Challenge {
            reason: StepUpChallengeReason::RequiredFactorMissing,
            assurance,
        });
    }
    if assurance.level < policy.minimum_level {
        return Ok(StepUpDecision::Challenge {
            reason: StepUpChallengeReason::InsufficientAssurance,
            assurance,
        });
    }

    if let Some(max_age_ms) = policy.max_age_ms {
        let max_age_ms = i64::try_from(max_age_ms)
            .map_err(|_| AuthError::ArithmeticOverflow("max_age_ms exceeds i64"))?;
        let cutoff_ms = now_ms
            .checked_sub(max_age_ms)
            .ok_or(AuthError::ArithmeticOverflow(
                "step-up cutoff overflowed i64",
            ))?;
        let assurance_is_stale = assurance
            .authenticated_at_ms
            .is_none_or(|authenticated_at_ms| authenticated_at_ms < cutoff_ms);
        let required_factor_is_stale = policy.required_factors.iter().any(|factor| {
            latest[factor_index(*factor)].is_none_or(|verified_at_ms| verified_at_ms < cutoff_ms)
        });
        if assurance_is_stale || required_factor_is_stale {
            return Ok(StepUpDecision::Challenge {
                reason: StepUpChallengeReason::Stale,
                assurance,
            });
        }
    }

    Ok(StepUpDecision::Allow { assurance })
}

/// Selects the first enrolled challenge factor in application-supplied preference order.
#[must_use]
pub fn select_mfa_challenge_factor(
    enrolled_factors: &[MfaChallengeFactor],
    preferred_factors: &[MfaChallengeFactor],
) -> Option<MfaChallengeFactor> {
    preferred_factors
        .iter()
        .find(|factor| enrolled_factors.contains(factor))
        .copied()
}

/// Evaluates already-decoded MFA challenge claims without owning token cryptography.
///
/// # Errors
///
/// Returns [`AuthError::InvalidValue`] when a policy has no allowed purpose or factor, or when an
/// expiration timestamp is supplied without the current time.
pub fn evaluate_mfa_challenge(
    facts: MfaChallengeFacts<'_>,
    policy: MfaChallengePolicy<'_>,
) -> AuthResult<MfaChallengeDecision> {
    if policy.allowed_purposes.is_empty() || policy.allowed_factors.is_empty() {
        return Err(AuthError::InvalidValue(
            "MFA challenge policy must allow a purpose and factor",
        ));
    }
    let (Some(purpose), Some(subject), Some(factor), Some(credential_version)) = (
        facts.purpose,
        facts.subject,
        facts.factor,
        facts.credential_version,
    ) else {
        return Ok(MfaChallengeDecision::Reject(
            MfaChallengeRejectionReason::Malformed,
        ));
    };
    if purpose.is_empty() || subject.is_empty() {
        return Ok(MfaChallengeDecision::Reject(
            MfaChallengeRejectionReason::Malformed,
        ));
    }
    if !policy.allowed_purposes.contains(&purpose) {
        return Ok(MfaChallengeDecision::Reject(
            MfaChallengeRejectionReason::PurposeMismatch,
        ));
    }
    if !policy.allowed_factors.contains(&factor) {
        return Ok(MfaChallengeDecision::Reject(
            MfaChallengeRejectionReason::FactorNotAllowed,
        ));
    }
    if policy
        .expected_subject
        .is_some_and(|expected| subject != expected)
    {
        return Ok(MfaChallengeDecision::Reject(
            MfaChallengeRejectionReason::SubjectMismatch,
        ));
    }
    if policy
        .current_credential_version
        .is_some_and(|current| credential_version != current)
    {
        return Ok(MfaChallengeDecision::Reject(
            MfaChallengeRejectionReason::CredentialVersionMismatch,
        ));
    }
    if facts.consumed_at_ms.is_some() {
        return Ok(MfaChallengeDecision::Reject(
            MfaChallengeRejectionReason::AlreadyConsumed,
        ));
    }
    if let Some(expires_at_ms) = facts.expires_at_ms {
        let Some(now_ms) = policy.now_ms else {
            return Err(AuthError::InvalidValue(
                "now_ms is required when expires_at_ms is present",
            ));
        };
        if now_ms >= expires_at_ms {
            return Ok(MfaChallengeDecision::Reject(
                MfaChallengeRejectionReason::Expired,
            ));
        }
    }

    Ok(MfaChallengeDecision::Accept)
}
